In [1]:
import pandas as pd
import numpy as np

credits = pd.read_csv("../dataset/tmdb_5000_credits.csv")
movies= pd.read_csv("../dataset/tmdb_5000_movies.csv")

In [2]:
credits.head(1)

,movie_id,title,cast,crew
0,19995,Avatar,"[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."


In [3]:
print(credits.head(1)['crew'].values)

<ArrowStringArray>
['[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name

In [4]:
movies.head(1)

,budget,genres,homepage,id,keywords,original_language,original_title,overview,popularity,production_companies,production_countries,release_date,revenue,runtime,spoken_languages,status,tagline,title,vote_average,vote_count
0,237000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.avatarmovie.com/,19995,"[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...",en,Avatar,"In the 22nd century, a paraplegic Marine is di...",150.437577,"[{""name"": ""Ingenious Film Partners"", ""id"": 289...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2009-12-10,2787965087,162.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,Enter the World of Pandora.,Avatar,7.2,11800


In [5]:
#movies.shape ## Rows=4803 and column = 20##Now it increased because 

In [6]:
#credits.shape

Merging both the data....because 2 dataset of similar kind is hectic

In [7]:
movies=movies.merge(credits,on='title')#.shape##Rows increased because of dulpicates data and column =23 because title id is one##NEVER RUN THIS COLUMN 2 TIMES BECAUSE IT DO THE FUNCTION TWICE AND DUPLICATES WILL HAPPEN

In [8]:
movies.head(1)

,budget,genres,homepage,id,keywords,original_language,original_title,overview,popularity,production_companies,...,runtime,spoken_languages,status,tagline,title,vote_average,vote_count,movie_id,cast,crew
0,237000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.avatarmovie.com/,19995,"[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...",en,Avatar,"In the 22nd century, a paraplegic Marine is di...",150.437577,"[{""name"": ""Ingenious Film Partners"", ""id"": 289...",...,162.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,Enter the World of Pandora.,Avatar,7.2,11800,19995,"[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."


Data Cleaning And Preprocessing

In [9]:
#Now will decide which column we should keep for our content based recommender system
#genres
#id
#keywords
#title
#overview
#cast
#crew

movies=movies[['movie_id','title','overview',
               'genres','keywords','cast','crew']]


In [10]:
#movies['original_language'].value_counts()

In [11]:
#movies.info()

In [12]:
#movies['tagline']

In [13]:
movies.head(1)

,movie_id,title,overview,genres,keywords,cast,crew
0,19995,Avatar,"In the 22nd century, a paraplegic Marine is di...","[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...","[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...","[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."


In [14]:
movies.isnull().sum()

movie_id    0
title       0
overview    3
genres      0
keywords    0
cast        0
crew        0
dtype: int64

In [15]:
movies.dropna(inplace=True)##It will remove the missing value row 
#We do this because missing value in overview column is 3 so doesnot affect our data

In [16]:
movies.duplicated().sum()##To check we have duplicated value or not

np.int64(0)

In [17]:
movies.iloc[0].genres

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

In [18]:
#we have to convert this [{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]' to
## Action,Adventure,fantasy,Scifi

In [19]:
import ast # def function will not work because geners data is present in string'' so we have to convert it in list of dic so for that we use this ast
           ###Data that are come from csvCSV,DATABASE,API JSON are stored in string so we should check and convert that in list [if need for the model]


In [20]:
##This function will help to fetch name
def convert(obj):
    L=[]
    for i in ast.literal_eval(obj):
      L.append(i['name'])
    return L

In [21]:
movies['genres']=movies['genres'].apply(convert) #Now my genres coloum is ready

In [22]:
movies.head(1)

,movie_id,title,overview,genres,keywords,cast,crew
0,19995,Avatar,"In the 22nd century, a paraplegic Marine is di...","[Action, Adventure, Fantasy, Science Fiction]","[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...","[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."


In [23]:
#movies.iloc[0]['cast']

In [24]:
#movies['keywords'] Same for keyword also

In [25]:
movies['keywords']=movies['keywords'].apply(convert)#Now keyword is ready

In [26]:
movies.head(5)

,movie_id,title,overview,genres,keywords,cast,crew
0,19995,Avatar,"In the 22nd century, a paraplegic Marine is di...","[Action, Adventure, Fantasy, Science Fiction]","[culture clash, future, space war, space colon...","[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."
1,285,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...","[Adventure, Fantasy, Action]","[ocean, drug abuse, exotic island, east india ...","[{""cast_id"": 4, ""character"": ""Captain Jack Spa...","[{""credit_id"": ""52fe4232c3a36847f800b579"", ""de..."
2,206647,Spectre,A cryptic message from Bond’s past sends him o...,"[Action, Adventure, Crime]","[spy, based on novel, secret agent, sequel, mi...","[{""cast_id"": 1, ""character"": ""James Bond"", ""cr...","[{""credit_id"": ""54805967c3a36829b5002c41"", ""de..."
3,49026,The Dark Knight Rises,Following the death of District Attorney Harve...,"[Action, Crime, Drama, Thriller]","[dc comics, crime fighter, terrorist, secret i...","[{""cast_id"": 2, ""character"": ""Bruce Wayne / Ba...","[{""credit_id"": ""52fe4781c3a36847f81398c3"", ""de..."
4,49529,John Carter,"John Carter is a war-weary, former military ca...","[Action, Adventure, Science Fiction]","[based on novel, mars, medallion, space travel...","[{""cast_id"": 5, ""character"": ""John Carter"", ""c...","[{""credit_id"": ""52fe479ac3a36847f813eaa3"", ""de..."


In [27]:
movies['cast']#we want only 3 name from the cast 

0       [{"cast_id": 242, "character": "Jake Sully", "...
1       [{"cast_id": 4, "character": "Captain Jack Spa...
2       [{"cast_id": 1, "character": "James Bond", "cr...
3       [{"cast_id": 2, "character": "Bruce Wayne / Ba...
4       [{"cast_id": 5, "character": "John Carter", "c...
                              ...                        
4804    [{"cast_id": 1, "character": "El Mariachi", "c...
4805    [{"cast_id": 1, "character": "Buzzy", "credit_...
4806    [{"cast_id": 8, "character": "Oliver O\u2019To...
4807    [{"cast_id": 3, "character": "Sam", "credit_id...
4808    [{"cast_id": 3, "character": "Herself", "credi...
Name: cast, Length: 4806, dtype: str

In [28]:
##This function is used to get first 3 name
def convert2(obj):
    L=[]
    counter=0
    for i in ast.literal_eval(obj):
        if counter != 3:
          L.append(i['name'])
          counter+=1
        else:
            break
      
    return L

In [29]:
movies['cast']=movies['cast'].apply(convert2)#Cast is ready

In [30]:
movies.head()

,movie_id,title,overview,genres,keywords,cast,crew
0,19995,Avatar,"In the 22nd century, a paraplegic Marine is di...","[Action, Adventure, Fantasy, Science Fiction]","[culture clash, future, space war, space colon...","[Sam Worthington, Zoe Saldana, Sigourney Weaver]","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."
1,285,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...","[Adventure, Fantasy, Action]","[ocean, drug abuse, exotic island, east india ...","[Johnny Depp, Orlando Bloom, Keira Knightley]","[{""credit_id"": ""52fe4232c3a36847f800b579"", ""de..."
2,206647,Spectre,A cryptic message from Bond’s past sends him o...,"[Action, Adventure, Crime]","[spy, based on novel, secret agent, sequel, mi...","[Daniel Craig, Christoph Waltz, Léa Seydoux]","[{""credit_id"": ""54805967c3a36829b5002c41"", ""de..."
3,49026,The Dark Knight Rises,Following the death of District Attorney Harve...,"[Action, Crime, Drama, Thriller]","[dc comics, crime fighter, terrorist, secret i...","[Christian Bale, Michael Caine, Gary Oldman]","[{""credit_id"": ""52fe4781c3a36847f81398c3"", ""de..."
4,49529,John Carter,"John Carter is a war-weary, former military ca...","[Action, Adventure, Science Fiction]","[based on novel, mars, medallion, space travel...","[Taylor Kitsch, Lynn Collins, Samantha Morton]","[{""credit_id"": ""52fe479ac3a36847f813eaa3"", ""de..."


In [31]:
movies['crew'][0]##Many crew member are there in movies but we only want director

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

In [32]:
##Function for fetching director name 
def fetch_director(obj):
    L=[]
    for i in ast.literal_eval(obj):
        if i['job']=='Director':
         L.append(i['name'])
         break
    return L

In [33]:
movies['crew']=movies['crew'].apply(fetch_director)##Crew is Ready

In [34]:
movies.head(5)

,movie_id,title,overview,genres,keywords,cast,crew
0,19995,Avatar,"In the 22nd century, a paraplegic Marine is di...","[Action, Adventure, Fantasy, Science Fiction]","[culture clash, future, space war, space colon...","[Sam Worthington, Zoe Saldana, Sigourney Weaver]",[James Cameron]
1,285,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...","[Adventure, Fantasy, Action]","[ocean, drug abuse, exotic island, east india ...","[Johnny Depp, Orlando Bloom, Keira Knightley]",[Gore Verbinski]
2,206647,Spectre,A cryptic message from Bond’s past sends him o...,"[Action, Adventure, Crime]","[spy, based on novel, secret agent, sequel, mi...","[Daniel Craig, Christoph Waltz, Léa Seydoux]",[Sam Mendes]
3,49026,The Dark Knight Rises,Following the death of District Attorney Harve...,"[Action, Crime, Drama, Thriller]","[dc comics, crime fighter, terrorist, secret i...","[Christian Bale, Michael Caine, Gary Oldman]",[Christopher Nolan]
4,49529,John Carter,"John Carter is a war-weary, former military ca...","[Action, Adventure, Science Fiction]","[based on novel, mars, medallion, space travel...","[Taylor Kitsch, Lynn Collins, Samantha Morton]",[Andrew Stanton]


In [35]:
type(movies['cast'])

pandas.Series

In [36]:
type(movies.iloc[0].overview)

str

In [37]:
movies.iloc[0]['overview']#Should tokenization(So we convert sentence into LIST OF WORDS.)

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.'

In [38]:
movies['overview']=movies['overview'].apply(lambda x:x.split())#Overview is ready

In [39]:
##Lambda x: is a small anonymous function.x=each overview sentence
##split()use to break sentence into words using spaces


In [40]:
movies['overview']

0       [In, the, 22nd, century,, a, paraplegic, Marin...
1       [Captain, Barbossa,, long, believed, to, be, d...
2       [A, cryptic, message, from, Bond’s, past, send...
3       [Following, the, death, of, District, Attorney...
4       [John, Carter, is, a, war-weary,, former, mili...
                              ...                        
4804    [El, Mariachi, just, wants, to, play, his, gui...
4805    [A, newlywed, couple's, honeymoon, is, upended...
4806    ["Signed,, Sealed,, Delivered", introduces, a,...
4807    [When, ambitious, New, York, attorney, Sam, is...
4808    [Ever, since, the, second, grade, when, he, fi...
Name: overview, Length: 4806, dtype: object

In [41]:
movies.head()

,movie_id,title,overview,genres,keywords,cast,crew
0,19995,Avatar,"[In, the, 22nd, century,, a, paraplegic, Marin...","[Action, Adventure, Fantasy, Science Fiction]","[culture clash, future, space war, space colon...","[Sam Worthington, Zoe Saldana, Sigourney Weaver]",[James Cameron]
1,285,Pirates of the Caribbean: At World's End,"[Captain, Barbossa,, long, believed, to, be, d...","[Adventure, Fantasy, Action]","[ocean, drug abuse, exotic island, east india ...","[Johnny Depp, Orlando Bloom, Keira Knightley]",[Gore Verbinski]
2,206647,Spectre,"[A, cryptic, message, from, Bond’s, past, send...","[Action, Adventure, Crime]","[spy, based on novel, secret agent, sequel, mi...","[Daniel Craig, Christoph Waltz, Léa Seydoux]",[Sam Mendes]
3,49026,The Dark Knight Rises,"[Following, the, death, of, District, Attorney...","[Action, Crime, Drama, Thriller]","[dc comics, crime fighter, terrorist, secret i...","[Christian Bale, Michael Caine, Gary Oldman]",[Christopher Nolan]
4,49529,John Carter,"[John, Carter, is, a, war-weary,, former, mili...","[Action, Adventure, Science Fiction]","[based on novel, mars, medallion, space travel...","[Taylor Kitsch, Lynn Collins, Samantha Morton]",[Andrew Stanton]


In [42]:
##Now the column is ready But in generes,keywords,cast and crew
#We have to remove space sam worthington to samwarthing 
##because for better ml training

In [43]:
movies['genres']=movies['genres'].apply(lambda x:[i.replace(" ","")for i in x])
movies['keywords']=movies['keywords'].apply(lambda x:[i.replace(" ","")for i in x])
movies['cast']=movies['cast'].apply(lambda x:[i.replace(" ","")for i in x])
movies['crew']=movies['crew'].apply(lambda x:[i.replace(" ","")for i in x])

In [44]:
movies.head(1)

,movie_id,title,overview,genres,keywords,cast,crew
0,19995,Avatar,"[In, the, 22nd, century,, a, paraplegic, Marin...","[Action, Adventure, Fantasy, ScienceFiction]","[cultureclash, future, spacewar, spacecolony, ...","[SamWorthington, ZoeSaldana, SigourneyWeaver]",[JamesCameron]


Combine all the column into one 

In [45]:
movies['tags']=movies['overview'] + movies['genres'] + movies['keywords'] + movies['cast'] + movies['crew']

In [46]:
movies.head(2)

,movie_id,title,overview,genres,keywords,cast,crew,tags
0,19995,Avatar,"[In, the, 22nd, century,, a, paraplegic, Marin...","[Action, Adventure, Fantasy, ScienceFiction]","[cultureclash, future, spacewar, spacecolony, ...","[SamWorthington, ZoeSaldana, SigourneyWeaver]",[JamesCameron],"[In, the, 22nd, century,, a, paraplegic, Marin..."
1,285,Pirates of the Caribbean: At World's End,"[Captain, Barbossa,, long, believed, to, be, d...","[Adventure, Fantasy, Action]","[ocean, drugabuse, exoticisland, eastindiatrad...","[JohnnyDepp, OrlandoBloom, KeiraKnightley]",[GoreVerbinski],"[Captain, Barbossa,, long, believed, to, be, d..."


In [47]:
new_df=movies[['movie_id','title','tags']]

In [48]:
new_df.head()

,movie_id,title,tags
0,19995,Avatar,"[In, the, 22nd, century,, a, paraplegic, Marin..."
1,285,Pirates of the Caribbean: At World's End,"[Captain, Barbossa,, long, believed, to, be, d..."
2,206647,Spectre,"[A, cryptic, message, from, Bond’s, past, send..."
3,49026,The Dark Knight Rises,"[Following, the, death, of, District, Attorney..."
4,49529,John Carter,"[John, Carter, is, a, war-weary,, former, mili..."


In [49]:
##Now Tags in the form of list but we need to convert this list.
##Into string because for vectorization
##Why we need that? Vectorization meaning 
# ["convert the text into number because ML works on numbers only"]
##Now Why we convert this list into string because NLP libaries expect TEXT documents
##Libraries like:[CountVectorizer] were designed for 
# #articles
#documents
#paragraphs
#text
#NOT Python lists.

In [50]:
new_df['tags']=new_df['tags'].apply(lambda x:" ".join(x))

In [51]:
new_df.iloc[0]['tags']

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. Action Adventure Fantasy ScienceFiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d SamWorthington ZoeSaldana SigourneyWeaver JamesCameron'

In [52]:
new_df['tags']=new_df['tags'].apply(lambda x:x.lower())

In [53]:
new_df.tail()

,movie_id,title,tags
4804,9367,El Mariachi,el mariachi just wants to play his guitar and ...
4805,72766,Newlyweds,a newlywed couple's honeymoon is upended by th...
4806,231617,"Signed, Sealed, Delivered","""signed, sealed, delivered"" introduces a dedic..."
4807,126186,Shanghai Calling,when ambitious new york attorney sam is sent t...
4808,25975,My Date with Drew,ever since the second grade when he first saw ...


In [54]:

# we will convert the tags in vector..
# So no of movies=No of vector...If someone select one vector means one movie
# Closest 5vector will be recommended..for that particular movie 
# Using BAG OF WORDS to convert text to Number 
# Step 1..We will Merge all the tags into one Big_tags means
# Big_tags=tags1+tags2+tags3.....tags4808
# Step 2..From the Big_tags we will take 5000 most common words (Remove stopword (english) means(is,are...etc))
## And why we are taking 5000 words because less words is better
# Suppose we have 10000000 unique words using all words...very slow,memory heavy,noisy so we limit vocabulary
# Step 3..Now from those 5000 common words will we see how many tags words are common from Big_tags words
# Step 4. Then the recommender system works##

In [55]:
new_df['tags'][1]

"captain barbossa, long believed to be dead, has come back to life and is headed to the edge of the earth with will turner and elizabeth swann. but nothing is quite as it seems. adventure fantasy action ocean drugabuse exoticisland eastindiatradingcompany loveofone'slife traitor shipwreck strongwoman ship alliance calypso afterlife fighter pirate swashbuckler aftercreditsstinger johnnydepp orlandobloom keiraknightley goreverbinski"

In [56]:
''' WE will do stemming

What is Stemming? (Very Easy Explanation)

Stemming is a technique used in NLP (Natural Language Processing) to convert similar words into one common base word.

Simple Definition
Stemming

Removing the ending part of words to get the root/base form.

Example

Words:

playing
played
plays
player

After stemming:

play

All become similar root word.And the meaning is same only'''

' WE will do stemming\n\nWhat is Stemming? (Very Easy Explanation)\n\nStemming is a technique used in NLP (Natural Language Processing) to convert similar words into one common base word.\n\nSimple Definition\nStemming\n\nRemoving the ending part of words to get the root/base form.\n\nExample\n\nWords:\n\nplaying\nplayed\nplays\nplayer\n\nAfter stemming:\n\nplay\n\nAll become similar root word.And the meaning is same only'

In [57]:
pip install nltk

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.2 -> 26.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [58]:
from nltk.stem.porter import PorterStemmer
ps=PorterStemmer()

In [59]:
def stem(text):
    y=[]
    
    for i in text.split():
        y.append(ps.stem(i))
        
    return " ".join(y)

In [60]:
new_df['tags']=new_df['tags'].apply(stem)

In [61]:
from sklearn.feature_extraction.text import CountVectorizer
cv=CountVectorizer(max_features=5000,stop_words='english')

In [62]:
vectors=cv.fit_transform(new_df['tags']).toarray()

**What is CountVectorizer?

CountVectorizer is a tool used in Natural Language Processing 

(NLP).

It converts text (sentences, movie tags, reviews, etc.) into 

numbers so that machine learning models can understand the

 text.

Because ML models cannot understand words directly — they 

only understand numbers.

 **max_features=5000

This means:

Keep only the top 5000 most frequently occurring words.

Why?

Because datasets may contain thousands or millions of words.

Using all words:

increases memory usage

slows training

adds unnecessary noise

So we keep only important frequent words.

**stop_words='english'

means remove word like 'is' 'are' etc

**fit_transform()

This does TWO jobs:

fit()

Learns all unique important words from the dataset

transform()

Converts each sentence into numerical vector

** Sparse Matrix

Before .toarray():

fit_transform() returns a sparse matrix.

Why sparse?

Because most values are zeros.

Example:

[0 0 0 1 0 0 0 0 0 0]

Storing all zeros wastes memory.

So Scikit-learn stores only non-zero values internally.

This saves huge memory.

toarray()

.toarray()

Converts sparse matrix into normal NumPy array.

So now you can see:

array

([[0,1,0,1],

       [1,0,0,1]])


In [63]:
vectors[1]

array([0, 0, 0, ..., 0, 0, 0], shape=(5000,))

In [64]:
features=cv.get_feature_names_out()

In [65]:
features[:200]##First 200 word

array(['000', '007', '10', '100', '11', '12', '13', '14', '15', '16',
       '17', '17th', '18', '18th', '18thcenturi', '19', '1910', '1920',
       '1930', '1940', '1944', '1950', '1950s', '1960', '1960s', '1970',
       '1970s', '1971', '1974', '1976', '1980', '1985', '1990', '1999',
       '19th', '19thcenturi', '20', '200', '2003', '2009', '20th', '21st',
       '23', '24', '25', '30', '300', '3d', '40', '50', '500', '60', '70',
       '80', 'aaron', 'aaroneckhart', 'abandon', 'abduct',
       'abigailbreslin', 'abil', 'abl', 'aboard', 'abov', 'abus',
       'academ', 'academi', 'accept', 'access', 'accid', 'accident',
       'acclaim', 'accompani', 'accomplish', 'account', 'accus', 'ace',
       'achiev', 'acquaint', 'act', 'action', 'actionhero', 'activ',
       'activist', 'activities', 'actor', 'actress', 'actual', 'ad',
       'adam', 'adamsandl', 'adamshankman', 'adapt', 'add', 'addict',
       'adjust', 'admir', 'admit', 'adolesc', 'adopt', 'ador',
       'adrienbrodi', 'adu

In [66]:
'''JUST To Check The Data'''

'JUST To Check The Data'

In [67]:
type(new_df)

pandas.DataFrame

In [68]:
new_df.columns

Index(['movie_id', 'title', 'tags'], dtype='str')

In [69]:
vectors[0]

array([0, 0, 0, ..., 0, 0, 0], shape=(5000,))

In [70]:
new_df.head()

,movie_id,title,tags
0,19995,Avatar,"in the 22nd century, a parapleg marin is dispa..."
1,285,Pirates of the Caribbean: At World's End,"captain barbossa, long believ to be dead, ha c..."
2,206647,Spectre,a cryptic messag from bond’ past send him on a...
3,49026,The Dark Knight Rises,follow the death of district attorney harvey d...
4,49529,John Carter,"john carter is a war-weary, former militari ca..."


In [71]:
new_df['tags'][0]

'in the 22nd century, a parapleg marin is dispatch to the moon pandora on a uniqu mission, but becom torn between follow order and protect an alien civilization. action adventur fantasi sciencefict cultureclash futur spacewar spacecoloni societi spacetravel futurist romanc space alien tribe alienplanet cgi marin soldier battl loveaffair antiwar powerrel mindandsoul 3d samworthington zoesaldana sigourneyweav jamescameron'

In [72]:
vectors.shape

(4806, 5000)

In [73]:
cv.get_feature_names_out()[300:400] 


array(['arnold', 'arnoldschwarzenegg', 'arrang', 'arrangedmarriag',
       'arrest', 'arriv', 'arrog', 'art', 'arthur', 'artifact',
       'artifici', 'artificialintellig', 'artist', 'ash', 'ashley',
       'ashleyjudd', 'ashtonkutch', 'asia', 'asian', 'asid', 'ask',
       'aspect', 'aspir', 'assassin', 'assault', 'assembl', 'assign',
       'assist', 'assistant', 'associ', 'assum', 'asteroid', 'astronaut',
       'asylum', 'atheist', 'athlet', 'atom', 'atomicbomb', 'attack',
       'attacks', 'attempt', 'attempttoescap', 'attend', 'attent',
       'attic', 'attitud', 'attorney', 'attract', 'auction', 'audienc',
       'audit', 'august', 'aunt', 'austin', 'australia', 'australian',
       'author', 'autism', 'auto', 'automobilerac', 'aveng', 'averag',
       'avoid', 'await', 'awak', 'awaken', 'awar', 'award', 'away',
       'awkward', 'awri', 'awry', 'ax', 'babe', 'babi', 'baby',
       'bachelor', 'backdrop', 'background', 'backpack', 'bad', 'bag',
       'bahama', 'bail', 'balanc',

In [74]:
'''over'''

'over'

So now we have 4806 movie = 4806 vector And each 5000 Number..
So now we have caluculate the distance between the vectors..less distance=more similarity ... N more distance = less similariry


'''Now we need a way to measure:

“How similar are these two vectors?”

That is where:

*Euclidean Distance
*Cosine Similarity/Distance

come in.
1. Euclidean Distance

This is normal geometric distance.

Like distance between two points.

Formula:d=\sqrt{(x_2-x_1)^2+(y_2-y_1)^2}
Imagine 2 points:

one point = Movie A
another point = Movie B

Euclidean distance measures:

actual straight-line distance between them.
Movie vectors:

A = [1,2]
B = [4,6]

Distance:

√((4-1)² + (6-2)²)
= √(9 + 16)
= √25
= 5

So:

smaller distance → more similar
larger distance → less similar

Now why we are not using Euclidean here..let's see
Movie A

Tags:

action hero fight

After vectorization:

[1,1,1]
Movie B

Tags:

action action action hero fight

After vectorization:

[3,1,1]

Important observation

Both movies are about:

✅ action
✅ hero
✅ fight

So semantically:

THEY ARE SIMILAR MOVIES.

Correct?

YES.

Now what Euclidean Distance sees

Euclidean looks at:

actual numerical distance
Distance between:

[1,1,1]
[3,1,1]

becomes:

√((3-1)² + (1-1)² + (1-1)²)

= √4
= 2

Euclidean says:

“These are somewhat far.”

But actually they are similar movies.

Problem happens because Movie B repeated "action" more times.

Real issue

Euclidean cares about:

❌ magnitude (size)

Movie B became “larger vector”.

So Euclidean thinks:

bigger vector = farther movie

But in NLP we usually DON'T care about vector size.

We care about:

✅ pattern of words
✅ direction of meaning

2. Cosine Similarity

Cosine similarity checks:

angle between vectors

NOT actual distance.

Formula

cos(θ)=
∣∣A∣∣ ∣∣B∣∣
A⋅B
	​


Where:

A⋅B = dot product
∣∣A∣∣ = vector length
Main intuition

Cosine similarity cares about:

✅ direction
NOT size.

One-line summary

Euclidean gets confused by repeated word counts and vector size.

Cosine focuses only on word pattern/direction, so it works better for movie/text similarity.'''

In [75]:
 from sklearn.metrics.pairwise import cosine_similarity
 

In [76]:
similarity=cosine_similarity(vectors)

In [77]:
similarity[0] #It is showing how much the 1st movie is similar to other 4805 movies
#Diagonal is always 1 because it will compare withself also

array([1.        , 0.08346223, 0.0860309 , ..., 0.04499213, 0.        ,
       0.        ], shape=(4806,))

In [78]:
def recommend(movie):
    movie_index=new_df[new_df['title'].str.lower()==movie.lower()].index[0]
    distances=similarity[movie_index]
    movie_list=sorted(list(enumerate(distances)),reverse=True,key=lambda x:x[1])[1:6]
    
    for i in movie_list:
        print(new_df.iloc[i[0]].title)

In [79]:
'''Here it is the main function which will fetch the 5 similar movie what user search....
movie_index=to known movie index no like what is the position of that particular movie user search..
we use str lower it will convert all the movie in lower case index[0]...first row that will match we show for ex xyz movie present in 12th position so it will show..
then distance= similarity[movie_index]  that xyz movie compare to all movie and show similarity value 
movie_list=we use (sorted) to arrange in asending order....enumerate will give similarity index(0,0.9  1,0.8 2,0.7) like that...
reserve is true means arrange that i descending order...key lambda x:x[1] sorted the similarity index not that index value that we are give using enumerate
[1:6]:why not [0:5] because 0 is always that paritcular movie...
then loop to get movie name one by one till 5
WORKING:when user search any movie....1st it will find the movie index... then the similarity....and  finally sorted similarity and give top 5 movie
'''

'Here it is the main function which will fetch the 5 similar movie what user search....\nmovie_index=to known movie index no like what is the position of that particular movie user search..\nwe use str lower it will convert all the movie in lower case index[0]...first row that will match we show for ex xyz movie present in 12th position so it will show..\nthen distance= similarity[movie_index]  that xyz movie compare to all movie and show similarity value \nmovie_list=we use (sorted) to arrange in asending order....enumerate will give similarity index(0,0.9  1,0.8 2,0.7) like that...\nreserve is true means arrange that i descending order...key lambda x:x[1] sorted the similarity index not that index value that we are give using enumerate\n[1:6]:why not [0:5] because 0 is always that paritcular movie...\nthen loop to get movie name one by one till 5\nWORKING:when user search any movie....1st it will find the movie index... then the similarity....and  finally sorted similarity and give t

In [80]:
recommend('the little prince')##Done

Jonah: A VeggieTales Movie
The True Story of Puss 'n Boots
Silver Medalist
Puss in Boots
11:14


In [81]:
new_df.columns

Index(['movie_id', 'title', 'tags'], dtype='str')

In [82]:
new_df['title'][500:]

500                         2 Fast 2 Furious
501                        The Little Prince
502                             The Invasion
503     The Adventures of Rocky & Bullwinkle
504                  The Secret Life of Pets
                        ...                 
4804                             El Mariachi
4805                               Newlyweds
4806               Signed, Sealed, Delivered
4807                        Shanghai Calling
4808                       My Date with Drew
Name: title, Length: 4306, dtype: str

In [83]:
'''Inside notebook:
new_df
similarity
are created.
These are stored in:
RAM / memory
ONLY while notebook is running
If you close notebook:
everything disappears
Like:
new_df disappears
similarity disappears
ML work disappears from memory
Then How will app.py use them?
Your website file:
app.py
is separate.
It cannot access notebook memory directly.
So we SAVE the data into files.
That saving process is called:Pickle
Pickle does SAME thing
It saves Python objects into files.
Later you can reopen them.
pickle.dump():SAVE data into file
pickle.load():READ saved data back
'''

'Inside notebook:\nnew_df\nsimilarity\nare created.\nThese are stored in:\nRAM / memory\nONLY while notebook is running\nIf you close notebook:\neverything disappears\nLike:\nnew_df disappears\nsimilarity disappears\nML work disappears from memory\nThen How will app.py use them?\nYour website file:\napp.py\nis separate.\nIt cannot access notebook memory directly.\nSo we SAVE the data into files.\nThat saving process is called:Pickle\nPickle does SAME thing\nIt saves Python objects into files.\nLater you can reopen them.\npickle.dump():SAVE data into file\npickle.load():READ saved data back\n'

In [84]:
import pickle


In [85]:
pickle.dump(new_df,open('movie.plk','wb'))

In [86]:
'''
So we convert it to dic
Before saving, dataframe is converted into dictionary.
Why did he change?

Because dictionary is:

lighter
simpler
faster sometimes

for websites.'''#(now new_df is in dic form and in streamlit we will convert back into dataframe for the function like .value,title)

'\nSo we convert it to dic\nBefore saving, dataframe is converted into dictionary.\nWhy did he change?\n\nBecause dictionary is:\n\nlighter\nsimpler\nfaster sometimes\n\nfor websites.'

In [87]:
#pickle.dump(new_df.to_dict(),open('movie_dict.pkl','wb'))FOR UNDERSTANDING THAT NEW DF IS ALSO WORKING SO I USED THAT

In [88]:
pickle.dump(similarity,open('similarity.pkl','wb'))

In [89]:
new_df.values

array([[19995, 'Avatar',
        'in the 22nd century, a parapleg marin is dispatch to the moon pandora on a uniqu mission, but becom torn between follow order and protect an alien civilization. action adventur fantasi sciencefict cultureclash futur spacewar spacecoloni societi spacetravel futurist romanc space alien tribe alienplanet cgi marin soldier battl loveaffair antiwar powerrel mindandsoul 3d samworthington zoesaldana sigourneyweav jamescameron'],
       [285, "Pirates of the Caribbean: At World's End",
        "captain barbossa, long believ to be dead, ha come back to life and is head to the edg of the earth with will turner and elizabeth swann. but noth is quit as it seems. adventur fantasi action ocean drugabus exoticisland eastindiatradingcompani loveofone'slif traitor shipwreck strongwoman ship allianc calypso afterlif fighter pirat swashbuckl aftercreditssting johnnydepp orlandobloom keiraknightley goreverbinski"],
       [206647, 'Spectre',
        'a cryptic messag from

In [90]:
new_df['title']

0                                         Avatar
1       Pirates of the Caribbean: At World's End
2                                        Spectre
3                          The Dark Knight Rises
4                                    John Carter
                          ...                   
4804                                 El Mariachi
4805                                   Newlyweds
4806                   Signed, Sealed, Delivered
4807                            Shanghai Calling
4808                           My Date with Drew
Name: title, Length: 4806, dtype: str